src/include/rbd/librbd.h

```diff
@@ -403,6 +415,17 @@ CEPH_RBD_API int rbd_pool_metadata_list(rados_ioctx_t io_ctx, const char *start,
                                         size_t *key_len, char *values,
                                         size_t *vals_len);
 
+CEPH_RBD_API int rbd_config_pool_get(rados_ioctx_t io_ctx, const char *key,
+                                     char *value, size_t *val_len);
+CEPH_RBD_API int rbd_config_pool_set(rados_ioctx_t io_ctx, const char *key,
+                                     const char *value);
+CEPH_RBD_API int rbd_config_pool_remove(rados_ioctx_t io_ctx, const char *key);
+CEPH_RBD_API int rbd_config_pool_list(rados_ioctx_t io_ctx,
```

Comment:
The main motivation to add `rbd_config_pool_XYZ` was that if these were not available and a user would want to use image and pool metadata API methods, she would need to validate the config param herself, and for this she would need to call `rbd_config_list` first just to check that the key in the list. Actually it is not about optimization but to make users harder to mess config metadata.

If you don't think it is good enough reason then I will remove set/remove methods. I am not sure how I could use the same method `rbd_config_list` both for a pool and for a image. If you mean the same method name but different signatures (`rados_ioctx_t io_ctx` vs `rbd_image_t image`) then it will not work as they both are `void *`.

Reply:
If all the config API methods are removed and the CLI uses the image and pool metadata APIs, you wouldn't need any special API methods to list pool metadata. Of course, that probably also just implies that the ```rbd config image list``` method could also just list both pool and image metadata to determine the config override source, so perhaps we don't need any config-specific API methods (yet).

Reply:
In my current implementation (and it looks like it agrees with what you suggested initially) `rbd config image list` will return all config options, even those that does not have an image or pool override (source "config"). If we don't have a specialize `rbd_config_list` method the caller will not know how to filter metada list for only options that are valid and read config values for options that are not in metadata. Ok we can add this knowledge to `rbd` tool to make cli work (and hope we will not forget to update it when needed). But what about other API users? I can image the dashboard could want to list the config overrides...

Reply:
> But what about other API users? I can image the dashboard could want to list the config overrides...

Good point -- preventing duplication of complex logic between the dashboard and rbd CLI is important

Reply:
So, summarizing:
- `rbd_config_pool/image_list` methods remain (or please advise how I can do with the only `rbd_config_list` method). 
- The config `get/set/remove` API methods are to be removed and validation on set  is in metadata methods.
- The CLI will provide `rbd config pool/image get` method that will call `rbd_config_list` method internally.

Right?

Reply:
> - `rbd_config_pool/image_list` methods remain (or please advise how I can do with the only `rbd_config_list` method). 

Yes

> - The config `get/set/remove` API methods are to be removed and validation on set  is in metadata methods.

Yes

> - The CLI will provide `rbd config pool/image get` method that will call `rbd_config_list` method internally.

I was thinking it would just use the ```rbd_pool_metadata_get``` / ```rbd_metadata_get``` w/ the ```conf_``` prefix adding in by the CLI.

Reply:
> I was thinking it would just use the rbd_pool_metadata_get / rbd_metadata_get w/ the conf_ prefix adding in by the CLI.

And if they return ENOENT fallback to the config value? Just calling `rbd_config_list` and filtering the result looks simpler to me (I think we don't need to worry about optimization here).

Reply:
> And if they return ENOENT fallback to the config value? Just calling `rbd_config_list` and filtering the result looks simpler to me (I think we don't need to worry about optimization here).

No worries, I'm not trying to optimize it. I was just thinking that the `rbd config pool/image get` actions would fail w/ -ENOENT if you provided a key that isn't being overridden.

Reply:
Ok. I thought that falling back to the pool / config level would be more useful behavior from users point of view. But thinking more about this I am not sure, and now returning -ENOENT in this case is more appropriate. If you have not changed your mind I will update it as you suggested.